### PR TITLE
Introducing Raw Data tracking and Command Logs

### PR DESCRIPTION
**Pull request by gordonje**
A lot of interrelated changes in this PR, mostly inspired by #1189:

*    Introducing a scheme for tracking (in the app's database) release datetimes and sizes of CAL-ACCESS as RawDataVersions and the RawDataFiles included in each version's zip file (in the DATA/ directory)
*    For each RawDataFile, store the number of records and columns at each stage of transformation: download, clean and load. More specifically:
    *    `downloadcalaccessrawdata`'s prep() method inserts the RawDataFile records;
    *    `cleancalaccessrawfile` looks up the RawDataFile record and sets the download_records_count and download_column_count values (since this is where we were already calculating these values)
    *    `loadcalaccessrawfile` looks up the RawDataFile record and sets the clean_records_count and clean_column_count values (again, these values were already being calculated in this command)
    *    `verifycalaccessrawfile` looks up the RawDataFile and sets the load_records_count and load_column_count values (again...already calculating this in this command)
*    Introducing a scheme for logging start and finish datetimes for each execution of the `updatecalaccessrawdata`, `downloadcalaccessrawdata`, `cleancalaccessrawfile` and `loadcalaccessrawfile` commands
    *    Commands called by other commands (e.g., update usually calls the other three) refer back to their caller, which I believe lays the foundation for smarter resuming, that is, picking up where the user left off in the middle of cleaning or loading raw files (started discussing this in one of @bcipolli's PRs, #1153)
*    These new tracking models are defined in their own models .py file and do not inherit from CalAccessBaseModel, because they don't need those methods, and they are filtered out of the get_model_list() function, which is frequently used to iterate over the models that describe the actual raw data
*    Introducing a new command: `reportcalaccessrawdata`, which iterates over the data models, calling `verifycalaccessrawfile` for each. It then calculates total and percentages for cleaned records and loaded records, and prints these to the console, and saves a .CSV file of each files stats to the docs/ directory.
    *    Would like to do a little more work on this command before the next release. Specifically, I want to set up a template that will include other useful summary stats and will save that output, maybe also in the docs folder.

Caveats:
*    Executions of `updatecalaccessrawdata` where the `test-data` option is invoked will not be tracked because it isn't clear which version to assign to the tracking records. And probably we don't want these being mixed in with live data
*    Relatedly, if there isn't any RawDataVersion logged in the db, the `cleancalaccessrawfile`, `loadcalaccessrawfile` and `verifycalaccessrawfile` commands will not attempt to create CalAccessCommandLogs or update RawDataFile records. This was the solution to a late-stage issue exposed by our automated tests, which create new databases and work exclusively with test data. So once again, there's no obvious RawDataVersion to assign to the tracking records.
    *    This seemed a better alternative than adding a `test-data` to these three commands (similar to the one for `updatecalaccessrawdata`). I could be convinced otherwise.
    *    A more ideal solution might come out future work around archiving older RawDataVersions so that users could specify which CAL-ACCESS version they want to load, verify or whatever. Maybe this could happen as part digging further into dat.

Other notes:
*    I also added a start_datetime and duration() method to the CalAccessCommand base class for convenient printing of durations of commands. This might want to be more tightly integrated with the tracking records
*    I moved the get_download_metadata() along with the url attribute method from `downloadcalaccessrawdata` up to CalAccessCommand because at least `updatecalaccessrawdata` needs to check on the currently available raw data version, and other command's might want this as well.
*    I added a command_name attribute to CalAccessCommand in order to facilitate creation of CalAccessCommandLog records without hard-coding strings for the names of the commands. I was a little surprised that something like this wasn't built-in. If someone has a more straight-forward approach, I'm all ears
*    This code could definitely be more DRY than it is. It seems to me now we have two distinct kinds of commands: Those that act on one data file / model versus those that act on the entire database. We might consider sub-classing CalAccessCommand to make the distinctions more concrete. The specific need driving this for me was a desire to have a single place where we define methods for starting and stopping the CalAccessCommand log records, it isn't clear how to do that when those logs require file name value for some commands, but not others.

### CODE REVIEW

**Comment by bcipolli** [comment]
There's a lot here, so I will simply ask: do these changes still support multi-database projects?

**Comment by gordonje** [comment]
@bcipolli Yes, definitely. To be clear, though, the tracking models will only go in your default database. Does it make sense to use the `--database` option to determine where to store the tracking records? I think doing this will require adding that options to download, clean and verify.

**Comment by bcipolli** [comment]
I think that should be fine we use a django router to route models with the `calaccess_raw` app label to a different database. As long as the code doesn't explicitly set the database to something, and lets us use our router, then should work great!

**Comment by gordonje** [comment]
Yeah, what should happen is these new tracking records will end up in whichever db you have defined in Django settings as the default. 

Being able to store tracking records in whichever db the raw data was loaded seems like it might be appealing for users like you who are using multiple dbs. Am I totally wrong about that?

**Comment by gordonje** [comment]
Just remember another big thing to point out: The `--resume-download` option has been deprecated from `updatecalaccessrawdata`. Instead, this command will check if your previous update did not complete and if the version to which you started updating is still available, in which case it will automatically resume  your download (unless you skip it).

The direction I feel we are moving is to have resume be more automatic, with maybe a prompt to opt out and restart (maybe a `--force-restart` option too?). That's all likely to follow after we get this set of changes into master.

**Comment by bcipolli** [comment]
> Being able to store tracking records in whichever db the raw data was loaded seems like it might be appealing for users like you who are using multiple dbs. Am I totally wrong about that?

Not wrong at all... but it doesn't really matter. As long as you're being agnostic, it allows multi-db people to route as they wish--the router specifies the database to use, based on the model's `meta`(data).

If you specify to route to `default` database explicitly, then there could be a problem. The key line is here:
```python
        self.database = self.database or router.db_for_write(model=model)
```

So if a user doesn't specify a database, `self.database is None`, and the router is called to figure out where to put it. Django has a default router that will return `default` in the single-db case.

**Comment by palewire** [comment]
I'm also a fan of having have the tracking models open for redirection via the ``--database`` option and the router.

**Comment by aboutaaron** [comment]
> To be clear, though, the tracking models will only go in your default database. Does it make sense to use the --database option to determine where to store the tracking records?

Yeah this would be awesome in case where someone wants to store these results somewhere else. I could see this is a simple way to separate concerns when hacking on a project using this data

**Comment by gordonje** [comment]
With these latest commits, I've added multi-database support for the new tracking models. 

This had a little more sprawl than I anticipated. In a nutshell, because the log records have to reference a version and because the some of the commands create and update raw data file records, these records need to exist in the database in the command's scope. 

So I'm doing is copying those records from the default database when they are needed.

More specifically:

*    I've moved the `--database` option up to CalAccessCommand and added database attribute, which is the default database if the user didn't pass something in this option. And the update command passes the database option on to all the sub commands.
*    Also added query sets for RawDataVersion, RawDataFile and CalAccessCommandLog as attributes to CalAccessCommand to facilitate getting and creating these objects by the sub-class commands. And I'm manually setting the database via the using() methods.
*    Also added methods to get RawDataVersion and RawDataFile objects from the database in the command's scope or copy it from the default database.

There are a couple of other bugs I'm chasing this morning:
*    The finish_datetime isn't getting logged in the db for a handful of the clean commands.
*    The load command is throwing a `DataError: value too long for type character varying(9)` for the `CVR2_SO_CD` table, but I'm betting this is un-related to these changes and is, rather, caused by an actual increase to the possible character length of this field.

Other than that, I welcome more feedback. I know I also need to update the docs, but I would prefer to hold off on that until I know this is all about right.

File: calaccess_raw/management/commands/__init__.py [hunk 1]
```diff
@@ -20,6 +45,109 @@ def handle(self, *args, **options):
...
+            self.command_name = sys.argv[1]
+        else:
+            # or take the end of the command's full module name
+            self.command_name = sub(r'(.+\.)*', '', self.__class__.__module__)
+
+        self.database = options["database"] or 'default'
+
+        self.raw_data_versions = RawDataVersion.objects.using(self.database)
+        self.raw_data_files = RawDataFile.objects.using(self.database)
+        self.command_logs = CalAccessCommandLog.objects.using(self.database)
```
Comment (bcipolli) [inline calaccess_raw/management/commands/__init__.py]: If you don't specify a database, it should use the default for these objects. So I think you can get away with:

```python
    if self.database:
        self.raw_data_versions = RawDataVersion.objects.using(self.database)
        self.raw_data_files = RawDataFile.objects.using(self.database)
        self.command_logs = CalAccessCommandLog.objects.using(self.database)
```

File: calaccess_raw/management/commands/__init__.py [hunk 2]
```diff
@@ -20,6 +45,109 @@ def handle(self, *args, **options):
...
+        # in (edge) case when update is not called from command line
+        # if called from command line
+        if self._called_from_command_line:
+            # use the arg passed to manage.py as command name
+            self.command_name = sys.argv[1]
+        else:
+            # or take the end of the command's full module name
+            self.command_name = sub(r'(.+\.)*', '', self.__class__.__module__)
+
+        self.database = options["database"] or 'default'
```
Comment (bcipolli) [inline calaccess_raw/management/commands/__init__.py]: I think just leave `self.database = None` if it's not specified. The default is not `'default'`, but rather to allow Django to route the model. Release control :)

**Comment by bcipolli** [comment]
@gordonje Thanks for working on the multi-db support! I'm not sure what problem led you to set the default database for each model explicitly to `default`, rather than allowing the router to choose based on the model. 

Understanding that would help me advise; at first blush, it's not clear that it's the best approach to me.

File: calaccess_raw/management/commands/__init__.py (hunk 1 above)
Comment (gordonje) [inline calaccess_raw/management/commands/__init__.py]: We need to set up these query sets regardless of whether or not the user opts for a specific db. So could do:
```python
    if self.database:
        self.raw_data_versions = RawDataVersion.objects.using(self.database)
        self.raw_data_files = RawDataFile.objects.using(self.database)
        self.command_logs = CalAccessCommandLog.objects.using(self.database)
    else:
        self.raw_data_versions = RawDataVersion.objects
        self.raw_data_files = RawDataFile.objects.objects
        self.command_logs = CalAccessCommandLog.objects
```
But I'm pretty sure that, if the user wants to log records in db other than default, we have to manually set the db. Because the suggestion coming from db_for_write() router method for any of the tracking models will be `'default'`.

Glad we're talking about this more. Want to make sure we are on the same page about it.

Comment (bcipolli) [inline calaccess_raw/management/commands/__init__.py]: Ah, right, I didn't see the assignment there. Yes, that `else` statement looks good!

>I'm pretty sure that, if the user wants to log records in db other than default, we have to manually set the db.

Yep!

>The suggestion coming from db_for_write() router method for any of the tracking models will be 'default'.

That's true for the default router, but isn't true if you implement your own and point Django to your router in `settings.py`. E.g.

https://github.com/caciviclab/disclosure-backend/blob/master/disclosure/routers.py
https://github.com/caciviclab/disclosure-backend/blob/master/disclosure/settings.py#L96

With the router, I can pick the default database for any model. I do that to shove all `CalAccess` models into a `calaccess_raw` database, for example, to keep our table schema in our default database simpler.

**Comment by gordonje** [comment]
@bcipolli Yeah, I want to hash this out with you more to make sure we are on the same page.

The early snippet you posted from the load command:

```python
        self.database = self.database or router.db_for_write(model=model)
```
Sets the database for loading to either the user's option or the database suggest for the given model (which is always going to be default). For command's like download and update, there isn't really a specific model. But I suppose in calaccess_raw/management/commands/__init__.py we could do:

```python
        self.database = self.database or router.db_for_write(model=RawDataVersion)
```
Or one of the other tracking models. Maybe that actually is better than hard-coding the `'default'` string. But still need to set up these query sets for all the sub-class commands, regardless of whether the user opts for a specific db.

**Comment by bcipolli** [comment]
@gordonje Ok, I think the strategy from before is flawed.

```python
        self.database = self.database or router.db_for_write(model=model)
```
resets the command-wide database based on the router for a single model. That's a bug. **The database used for a model** should be `self.database or router.db_for_write(model=model)`, and `self.database` should stay as `None` if it starts as `None`.

That's my code, I just didn't quite get it right. The router should be called on each model one-by-one. I didn't catch the bug because all models are sent to the same db in my code, for the calaccess commands, and so setting it once was enough. But that's not correct.

So again, I think the `if...else` from above is correct, and that the old code has a bug.

**Comment by gordonje** [comment]
@bcipolli Actually, the `self.database` can't be `None` in `loadcalaccessrawfile` because we need to set up the connection and cursor and check which db engine is in use to decide whether to call `load_mysql()` or `load_postgresql()`). Ugh...

**Comment by bcipolli** [comment]
@gordonje That's fine then. We can deal with that later; let the assumption be that all related models go to the same database.

**Comment by gordonje** [comment]
Just to be clear what we want:

1.    If the user calls `loadcalaccessrawfile` **without** `--database`, both the loaded records and the RawDataVersion, RawDataFile and CalAccessCommandLog records should end up in the db suggested by the router for the given model.
2.    If the user calls `loadcalaccessrawfile` **with** `--database`, both the loaded records and the RawDataVersion, RawDataFile and CalAccessCommandLog records should end up in the database specified in the command option.

If so, then that's how I have it working in this last commit (tested with models that route to the default database or some other database). In `loadcalaccessrawfile` I overwrote the database and tracking record query set attributes. This might need to happen in `verifycalaccessrawfile` too, and so might be another argument for spinning CalAccessFileCommand (or CalAccessModelCommand?) off of CalAccessCommand.

**Comment by bcipolli** [comment]
@gordonje :+1: sounds good!

**Comment by palewire** [comment]
I've been out of town and haven't been reviewing this code very closely until today.

Looking at the code now, I see it aiming to support the ability to load the raw tables into separate databases with tracking tables synchronized across databases via a series of ``get_or_copy_`` methods, like [these](https://github.com/california-civic-data-coalition/django-calaccess-raw-data/blob/coverage-reports/calaccess_raw/management/commands/__init__.py#L83-L154). 

I wonder if that bit of complexity is necessary. What is the use case of this kind of thing and why isn't a higher-level switch without all that good enough?

**Comment by palewire** [comment]
This is probably more for my benefit that anyone else, but I would like to sketch out a couple use cases for the multi-database support here so we can clarify our needs and avoid making our code too complicated.

That's something I'm worried about when I see the complexity of the ``get_or_copy_`` methods discussed in my previous comment.

Here are the use cases I can imagine, _please correct me where I get them wrong and add more where you can see them_. I acknowledge my multi-database experience is slim and I likely have some misconceptions.

**Scenario 1: I don't want ``calaccess_raw`` in my default database**

Let's say, for instance, I am creating a public website that will be derived in part from CAL-ACCESS data. Some slices or extracts from it will be pulled out at a regular interval and reshaped for the live web, but none of the raw tables will be hit directly by HTTP requests. 

I could just drop all the raw tables into the database that also powers live site, but I'd prefer to keep it kosher and separate in a different database, or perhaps on an entirely different server. That way updates can be handled with separate processors and web traffic can be limited to the slimmed down set of tables that appear online. 

_My suspicion is that this is the most frequent and important use case for multi-database support and where we should focus our efforts._

To do this, I would install the ``calaccess_raw`` application from PyPI and add it to my ``INSTALLED_APPS`` per usual, but signal to Django when the tables are created that I would like them to be dropped in a separate database. I would do this by passing the option to the ``migrate`` command like so:

```bash
$ python manage.py migrate calaccess_raw --database=myseconddatabase
```

This would place all of the raw data tables as well as our tracking tables in the second database.

When updating the raw data via our management commands, I would not have to provide any extra configuration, but simply rely on ``router.db_for_write(model=self.model)`` to find and use the second database.

```bash
$ python manage.py updatecalaccessrawdata
```

**Scenario 2: I want to have more than one copy of the ``calaccess_raw`` tables that I update separately**

In this scenario I want to (for whatever reason) maintain multiple copies of the ``calaccess_raw`` database tables. I would create these copies by running the migrate command multiple times with different database options.

```bash
$ python manage.py migrate calaccess_raw --database=myseconddatabase
$ python manage.py migrate calaccess_raw --database=mythirddatabase
```

This would create copies of all raw tables and all tracking tables twice, with no direct link between the two.

When updating the database, I would provide a database option at the outset to point at which one I would like to work with.

```bash
$ python manage.py updatecalaccessrawdata --database=myseconddatabase
$ python manage.py updatecalaccessrawdata --database=mythirddatabase
```

This is the use case that I believe, because of a desire to avoid repeating the download and clean steps for each load, birthed the idea of either forcing the application to create only one set of tracking tables in the default database or, as the PR now does, synchronizing two sets of tracking tables between database 

However, unless we hear from @bcipolli and others that this is an important case that we should endeavor to optimize, I'm reluctant to introduce that kind of complexity to our code. Would we be doing anyone in the real world a disservice by forcing them to run parallel updates

**Comment by bcipolli** [comment]
@palewire I just want the first case--I want to keep `calaccess_raw` data out of my default database. I can't think of any practical scenario for the second one, at the moment.

**Comment by gordonje** [comment]
So in the first scenario, the none of the CalAccessCommands would need the `--database` option, right? That would definitely make things simpler.

**Comment by bcipolli** [comment]
Sounds that way. That works for well for me; I don't use the command-line arguments.

**Comment by palewire** [comment]
I believe that's right, @gordonje, but I wonder if it might be possible to simplify what you have now to continue to support second scenario by keeping the option but dropping the "copy" functionality.

**Comment by gordonje** [comment]
In scenario 2, a savvy user that really knows the interface and command options would reasonably believe "I ran the full update command once already and I used the keep files option, so since I now want to update my second and third database, I can just run update and skip the download and clean steps, and save some time.” After all, we are providing those exact options.

If they follow those exact steps, the update command would create a new version db record before all the calls to the load command, so we would be okay on that front. But ideally, I feel like we should only be creating new version records whenever the user downloads. Practically, though, the update command needs to set up it's log record, along with the associated version, before it calls the download command. 

So the update command currently is always creating new version records (if they don't already exist). But I wonder if it should be if the user skips the download step. What's implied here is that, when you call update, you're always updating to the version that's current available at the CAL-ACCESS url, even if you aren't actually downloading that version's files.

The sketchier scenario, which I thought we were trying to support and is also allowed by our interface, is when the user makes calls to load outside of the update command, and attempts to load some models into one database and some into another. The more I think about it, the more I realize how poorly we are handling that scenario and expecting that the user really knowing what they're doing because they would have to make sure that the write models are in the write database and all the tracking models are in all the databases.

But if all we expect is that some users just want to create full copies of the CAL-ACCESS database, wouldn't it be more straight-forward and faster to use the SQL commands like, in Postgres, `CREATE DATABASE [copy] WITH TEMPLATE [original]` or `mysqldump` in MySQL?

I'm not totally against maintaining the `--database`, but it starting to feel like there isn't a strong need for it, and removing it will definitely make things simpler. 

If we do decide to keep it, then we would probably want to add some more explicit exceptions to the load command to tell users when and why they need to run full updates. Also, with respect to the RawDataFile records that track which files are in which version and their record and column counts, we would need to allow the load command to make those records when it needs them. Don't have strong feelings about it, but I sorta don't like this because I had been thinking we should be notifying users if they are trying to load a file that has not yet been cleaned. But that's prob not that big of a deal.

**Comment by palewire** [comment]
This is a tough call, but I think I'm leaning towards the radical simplification suggested by @gordonje that would remove the ``--database`` option entirely and rely on the model inference of the router to determine the database to use. 

That, in turn, could be handled higher up the chain at the initial migration of the tables. Or by subclassing of the management commands.

This all of course banks on my psuedocode from Scenario 1 above actually working. But I'm finding the idea of simplification very appealing at this stage.

**Comment by gordonje** [comment]
The one thing that isn't totally explicit from @palewire's description of scenario 1 is that users will need to [implement Django router(s)](https://docs.djangoproject.com/en/1.9/topics/db/multi-db/#using-routers) if they want some or all of the calaccess_raw models in some database besides `default`. Django will not automatically assume the correct database for the model based on which the database option invoked when calling `migrate`.

I tested this by using our example project configuration with two databases configured, one aliased as `default` and the other as `alt`. Starting from two completely empty databases, I migrate only the `alt` db:

```bash
    python example/manage.py migrate --database=alt --run-syncdb
```

Then check to see what db is suggested for the tracking and campaign models:

```bash
    $ python example/manage.py shell
    >>> from django.db import router
    >>> from calaccess_raw.models.tracking import RawDataVersion
    >>> from calaccess_raw.models.campaign import CvrSoCd
    >>> router.db_for_write(RawDataVersion)
    'default'
    >>> router.db_for_write(CvrSoCd)
    'default'
```

If I then create a router.py file similar to the [one open-disclosure is using](https://github.com/bcipolli/opendisclosure-backend/blob/master/disclosure/routers.py) to route all models for the calaccess_raw app to the db aliased as `alt`. Then I add this line to my settings_local.py file:

```python
    DATABASE_ROUTERS = ['example.routers.ExampleRouter']
```

I get the following results:

```bash
    $ python example/manage.py shell
    >>> from django.db import router
    >>> from calaccess_raw.models.tracking import RawDataVersion
    >>> from calaccess_raw.models.campaign import CvrSoCd
    >>> router.db_for_write(RawDataVersion)
    'alt'
    >>> router.db_for_write(CvrSoCd)
    'alt'
```

tl;dr: If users want some or all of our models in some other database besides the one configured as the default for their Django project, the router system is the best way to do that. And it isn't clear why anyone would need the additional `--database` option on management commands to override this setting on an ad hoc basis.

I'm going go back over the code in this PR to scrub out the `--database` option and the `get_or_copy...`, and I'm going to add an explainer to our docs for users that want to use multiple dbs with links to the relevant sections of the Django docs.

**Comment by palewire** [comment]
Thank you for sorting all that out.

**Comment by gordonje** [comment]
Still want to add more to the output of the report, but tests are passing and I'm feeling confident the code is in a good place. 

Actually, it looks like the merge is going to be pretty clean. Unless anyone sees any glaring problems, would like to merge this PR so that isn't hanging around any longer. Of course, we can always make more improvements later.

**Comment by palewire** [comment]
Congratulations on working through this thorny problem. I'm so exciting I'm going to merge it right now!

**Comment by bcipolli** [comment]
Agreed, this is superb. Thanks so much for your time and care in working through this, @gordonje and @palewire !
